crates/fluvio-cli/src/bin/main.rs

```diff
@@ -27,10 +27,85 @@ fn print_help_hack() -> Result<()> {
         HelpOpt {}.process()?;
         std::process::exit(0);
     } else if let Some(first_arg) = args.nth(1) {
+        // We pick help up here as a courtesy
         if vec!["-h", "--help", "help"].contains(&first_arg.as_str()) {
             HelpOpt {}.process()?;
             std::process::exit(0);
         }
     }
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use clap::Parser;
+    use fluvio_cli::Root;
+
+    #[test]
+    fn test_correct_command_parsing_help() {
+        let should_not_succeed = vec![
+            "fluvio",
+            "fluvio -h",
+            "fluvio --help", // fluvio help is hacked with print_help_hack
+        ];
+        for s in should_not_succeed {
+            assert!(!parse_succeeds(s), "{s}");
+        }
+    }
+
+    #[test]
+    fn test_correct_command_parsing_consume() {
+        let should_succeed = vec![
+            "fluvio consume -H hello",
+            "fluvio consume -T hello",
+            "fluvio consume -H -n 10 hello",
+            "fluvio consume -T -n 10 hello",
+            "fluvio consume --start -n 0 hello",
+            "fluvio consume --start hello",
+            "fluvio consume hello --start --end 5",
+            "fluvio consume --start --end 5 -n 0 hello",
+        ];
+        for s in should_succeed {
+            assert!(parse_succeeds(s), "{s}");
+        }
+
+        let should_not_succeed = vec![
+            "fluvio consume",
+            "fluvio consume -H 0 hello",
+            "fluvio consume -T 0 hello",
+            "fluvio consume -H -n -10 hello",
+            "fluvio consume -H -n hello",
+            "fluvio consume -n 10 hello",
+            "fluvio consume -H -T -n 10 hello",
+            "fluvio consume -n hello",
+            "fluvio consume --start 5 hello",
+            "fluvio consume --end hello",
+        ];
+        for s in should_not_succeed {
+            assert!(!parse_succeeds(s), "{s}");
```

Comment:
There can be some gotchas with clap. For example, before this PR, running `target/../fluvio consume -B greetings` would fail. I wanted to make extra sure clap was working as expected.

Reply:
Why would that fail?  Works for me.

Reply:
Running both the master build and the installed cli I get
```
master ❯ flvd consume -B hello                                                                                                                                                                                                                                                
error: Invalid value 'hello' for '-B [<integer>]': invalid digit found in string                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                                                
For more information try '--help'            
```

Reply:
```
fluvio consume hello -B
Consuming records from the beginning of topic 'hello'
```⠄

Reply:
Yes. That is precisely the bug. Putting the flag before hello as opposed to after.

Reply:
`Usage: fluvio consume [OPTIONS] <topic>` https://www.fluvio.io/cli/commands/consume/

Reply:
Good catch!

Reply:
Thanks. This was the original bug that started this PR. Unfortunately, I think it means the docs might need to be updated as I don't think clap can handle the `Option<Option<T>>` case when there can than be arguments to the command as well.

Reply:
#### Responding to "There must be API to create options".

I did some digging into this. Options in and of themselves are straightforward. Nested options are the problem.

For clarity, below uses use the following precise definitions.
given the command `tail -n 5 file.txt`
`-n` is an `option`. (not to be confused with `Option` in rust).
`5` is a `parameter`
`file.txt` is an `argument`.

 The problem is that the previous usage had an optional `option` `-B` that had an optional `parameter`

The rust struct looked like this
``` rust
 // master branch
    pub struct ConsumeOpt {
        /// Consume records starting X from the beginning of the log (default: 0)
        #[clap(short = 'B', value_name = "integer", conflicts_with_all = &["offset", "tail"])]
        pub from_beginning: Option<Option<u32>>,
```

This gets applied greedily from left to right, so that running `fluvio -B topic` means that `topic` is parsed as the optional `parameter` to the  `option` `from_beginning` and not as the intended `argument`.

I couldn't find anyway to tell clap to prioritize an empty optional `parameter` in order to supply an `argument` to the command, presumably because the syntax is unavoidably ambiguous.

#### A potential alternative solution would be

1.  preempt the args array before passing to clap 
2.  if a required `argument` exists for the specified subcommand  then save the `argument` and truncate the array.
3.  parse potentially truncated array with clap
4. restore `argument` as needed.

This loses the elegance and convenience of clap though.

Reply:
The proposed solution was to make `from_beginning` a simple flag `option` and than provide another `option` `-n` that has a **required** `parameter` offset that provides the same behavior so that `fluvio -B -n 5` is equivalent to `fluvio -B 5` previously.